include/circt/LogicalEquivalence/Circuit.h

```diff
@@ -75,43 +84,148 @@ class Solver::Circuit {
   void performSub(mlir::Value result, mlir::OperandRange operands);
   void performXor(mlir::Value result, mlir::OperandRange operands);
 
+  // `seq` dialect operations.
+  void performCompReg(mlir::Value input, mlir::Value clk, mlir::Value data,
+                      mlir::Value reset, mlir::Value resetValue);
+  void performFirReg(mlir::Value next, mlir::Value clk, mlir::Value data,
+                     mlir::Value reset, mlir::Value resetValue);
+  void performFromClock(mlir::Value result, mlir::Value input);
+
+  // `verif` dialect operations.
+  void performAssert(mlir::Value property);
+  void performAssume(mlir::Value property);
+  void performCover(mlir::Value property);
+
 private:
-  /// Helper function for performing a variadic operation: it executes a lambda
-  /// over a range of operands.
-  void variadicOperation(
-      mlir::Value result, mlir::OperandRange operands,
-      llvm::function_ref<z3::expr(const z3::expr &, const z3::expr &)>
-          operation);
+  /// Struct to represent computational registers
+  struct CompRegStruct {
+    mlir::Value input;
+    mlir::Value clk;
+    mlir::Value data;
+    mlir::Value reset;
+    mlir::Value resetValue;
+  };
+  /// Struct to represent FIRTLL registers
+  struct FirRegStruct {
+    mlir::Value next;
+    mlir::Value clk;
+    mlir::Value data;
+    mlir::Value reset;
+    mlir::Value resetValue;
+    bool isAsync;
+  };
+  /// A type for the different operand sets a wire might need to store
+  using WireVariant = std::variant<
+      std::tuple<mlir::Value, llvm::StringLiteral>,
+      std::tuple<mlir::Value, mlir::Value, llvm::StringLiteral>,
+      std::tuple<mlir::Value, mlir::Value, mlir::Value, llvm::StringLiteral>,
+      /*Variadic Ops*/
+      std::tuple<mlir::OperandRange, llvm::StringLiteral>,
+      /*ICmpOp:*/
+      std::tuple<circt::comb::ICmpPredicate, mlir::Value, mlir::Value,
+                 llvm::StringLiteral>,
+      /*ParityOp & ReplicateOp:*/
+      std::tuple<mlir::Value, unsigned int, llvm::StringLiteral>,
+      /*ExtractOp:*/
+      std::tuple<mlir::Value, uint32_t, int, llvm::StringLiteral>>;
```

Comment:
I suspect the indirection through `Circuit` is causing this churn here. Might be worthwhile trying to delete that whole thing? Pretty sure that the hundreds of lines of overhead the indirection introduces could boil down to those 200-400 lines of `performX` calls directly inlined into a visitor, and then calling that from an `op.walk([&](Operation *op){ visitor.dispatch(op); });` call. WDYT?

The LogicExporter feels like it really wants to be something that walks the IR, ideally depth-first such that there are no `constrainResults` calls, taking in values for the inputs and current reg state, and producing new Z3 expressions for the outputs and next reg state. That's a bit annoying because of the depth-first search, so alternatively it could have four maps, with entries for the inputs, current reg values, outputs, and next reg values.

Reply:
Yeah, I think a lot of this is just bulk working around Circuit. Letting the LogicExporter get its hands on the Z3 directly seems way more elegant, and would prune a lot of the bulk out I think (although I do think the Circuit architecture was a nice separation of concerns for LEC. Just tricky when we need to do more with the IR)